format-specs/compatible-parquet.md

```diff
@@ -12,7 +12,7 @@ The core idea of the compatibility guidelines is to have the output match the de
 
 * The geometry column should be named either `"geometry"` or `"geography"`.
 
-* The geometry column should be a `BYTE_ARRAY` with Well Known Binary (WKB) used to define the geometries, as defined in the [encoding](./geoparquet.md#encoding) section of the GeoParquet spec.
+* The geometry column should be a `BYTE_ARRAY` with Well Known Binary (WKB) used to define the geometries, as defined in the [encoding](./geoparquet.md#encoding) section of the GeoParquet spec. Alternatively, the geometry column can be stored according to the Point, MultiPoint, MultiLineString, or MultiPolygon memory layouts with separated (struct) coordinates as specified in the [GeoArrow format](https://geoarrow.org/format).
```

Comment:
Does this mean that geometry column with mixed types of geometries cannot be encoded as GeoArrow?

Reply:
Not until https://github.com/apache/parquet-format/pull/44 is merged (if ever)

Reply:
Columns with mixed geometry values are quite common for most query engines with geospatial support. Most of the time geometry columns have the umbrella type "geometry" or "geography", and it is not practical to first resolve the subtypes of the geometries before writing out parquet files. I'd look forward to a columnar encoding supporting mixed geometry types as well as geometry collections.

Reply:
Arrow and Parquet are two different specs. Arrow has a union type, which allows for mixed geometry types in a single column, while maintaining constant-time access to any coordinate. Parquet does not today have a union type, so it's impossible to to write the `Geometry` and `GeometryCollection` arrays in https://github.com/geoarrow/geoarrow/pull/43 natively to Parquet.

GeoArrow implementations are able to statically know whether they have singly-typed geometries in a column, in which case they can write one of the 6 primitive types. GeoArrow implementations will have to fall back to WKB-encoded geometries for mixed-type columns. I don't see how this is something we could realistically change, unless we essentially [re-implement union handling in a struct](https://github.com/geoarrow/geoarrow/pull/43#pullrequestreview-1560131039), which would be a big ask for implementors.

Reply:
I understand that Arrow and Parquet are different specs, and it is impossible to write mixed-typed geometry values to Parquet files using GeoArrow encoding. So do we have to define yet another encoding and add it to the spec if we want to support encoding mixed-typed geometry values in Parquet in a columnar fashion (preserving all the benefits of row-group and page level statistics)? The new encoding could be something similar to `Struct<geometry_type: int8, multipolygon>` or [Spatial Parquet](https://arxiv.org/abs/2209.02158).

Reply:
> So do we have to define yet another encoding and add it to the spec if we want to support encoding mixed-typed geometry values in Parquet in a columnar fashion (preserving all the benefits of row-group and page level statistics)? 

No, we can have a WKB column plus a bounding box column (https://github.com/opengeospatial/geoparquet/pull/191).

> The new encoding could be something similar to `Struct<geometry_type: int8, multipolygon>`

This is what @paleolimbot offered in https://github.com/geoarrow/geoarrow/pull/43#pullrequestreview-1560131039 and what I referred to as "re-implement union handling in a struct". 

In any case, I'd argue that a large majority of use cases will be solved by these 6 primitive types, and we can come back to union types in the future.

Reply:
@paleolimbot and @kylebarron Thanks for leading the effort on this. This is great work.

I believe @Kontinuation 's point is more like: without allowing mixed geometry types in a column, all database engines that didn't use Arrow as the in-memory format have no way to know if a geometry column is a mixed type or not. In other words, they don't know if they can use this GeoArrow encoding to GeoParquet files.

The only way to know it is that scanning every single record of a big dataset first (get all geometry types), then in the second round, write the table to GeoParquet. This process will significant slow down the write performance and add challenges to the implementer or database users.

***So, regardless if the 6 primitive types cover the majority of the use cases or not, as long as a database engine wants to write GeoArrow encoding in GeoParquet, this additional check is inevitable.***

* Many database engines are gonna be affected by this: Snowflake, Big Query, AWS Redshift, Apache Sedona, Trino. Because their Geometry/Geography column allows mixed geometries.

* Engines will not be affected by this: DataFusion, GeoPandas (with Arrow), or any engine that uses Arrow/GeoArrow as its in-memory layout.

On the other hand, according to https://github.com/apache/parquet-format/pull/44, the Union type of Parquet might never be accepted by the Parquet community given that Apache Iceberg also does not allow Union type. Multiple Apache Iceberg PMCs are Apache Parquet PMCs as well. ***This means it is very likely that we will have to `re-implement union handling in a struct`***

That said, I think we can still add the GeoArrow encoding to the GeoParquet spec since I don't think this will create much incompatibility if we fix this issue in the future.

***The only downside of adding this to the spec now is that the adoption of writing GeoArrow encoding in GeoParquet might be low.***

Just my 2 cents. Please feel free to correct me if I am wrong.

Reply:
This discussion is making me think that the GeoParquet spec should not be defined _in terms of_ GeoArrow. Rather it should be defined as "native type" or "flat type" or similar. Then a sentence in prose can mention that it overlaps partially with the GeoArrow spec.

I'm also becoming convinced that the serialized form need not exactly overlap with GeoArrow. On the topic of mixed arrays specifically, as possibly the only one who has written an implementation of GeoArrow mixed arrays, I've become an even stronger proponent of using an Arrow union type for GeoArrow mixed arrays because of its ability to know geometry types statically. So I think the best way forward for GeoParquet (for a future PR) would be to discuss a "struct-union" approach for GeoParquet that is _not_ the same in-memory representation as GeoArrow.

> all database engines that didn't use Arrow as the in-memory format have no way to know if a geometry column is a mixed type or not. In other words, they don't know if they can use this GeoArrow encoding to GeoParquet files

I think changing nomenclature will also be clearer to non-arrow-based implementations that reading and writing this "native" encoding of GeoParquet is _not_ dependent on using Arrow internally.

So my recommendation would be to **take out most references to geoarrow** from this PR. I.e., we don't want the metadata key called `geoarrow_type` if there's a possibility where the GeoParquet union type is not the same as the GeoArrow union type.

Reply:
> This discussion is making me think that the GeoParquet spec should not be defined _in terms of_ GeoArrow. Rather it should be defined as "native type" or "flat type" or similar. Then a sentence in prose can mention that it overlaps partially with the GeoArrow spec.

We could go back to to the question about which "encoding" values to allow, and instead of a generic `"geoarrow"` option (with an additional "geoarrow_type" key to then be more specific), have actual encoding options `"point"`, `"linestring"`, `"polygon"`, etc.

For the rest, it is mostly a question about how to document this: how to phrase this exactly in the specification, how strongly to tie it to geoarrow (or just reference as mostly similar), how much to duplicate the details of the memory layout, etc. But that's all "details" about the best way to document it, while keeping the actual specification (what ends up in the metadata in a file) agnostic to geoarrow. 

I think I am somewhat convinced by @kylebarron's points on this, and like to idea of having the actual spec changes not use "geoarrow" (and then we can still debate how much to use the term in the explanation of it).

For example, as long as the specification would exactly overlap (or be a strict subset of geoarrow), we can still point to the geoarrow spec for the details to avoid too much duplication (and having to maintain two versions). And this is also easy to change then in the future if we would want to introduce differences.

>  I also don't think it would be productive to link to the GeoArrow specification for documentation of all the memory layouts but very pointedly not call it GeoArrow.

On the other hand, for an implementation of GeoParquet in some library that has nothing to do with Arrow (doesn't use an Arrow implementation under the hood), the "geoarrow" name is also somewhat uninformative, when strictly looking at it from a GeoParquet point of view.